### PR TITLE
component content

### PR DESCRIPTION
**Pull request by volodpj**
https://volodpj.github.io/todo-react/

### CODE REVIEW

**Comment by volodpj** [comment]
https://volodpj.github.io/todo-react/

File: src/App.js [hunk 1]
```diff
@@ -1,38 +1,118 @@
...
+class App extends React.Component  {
+    constructor(props) {
+        super(props);
+        this.state = {
+            todos: {
+               
+            }
+          }
+        
+        this.generateNewId = () => {
```
Comment (ivadimko) [inline src/App.js]: Suggestion: use `Date.now()` as id
Suggestion 2: use `random string` library: https://www.npmjs.com/package/randomstring

File: src/App.js [hunk 2]
```diff
@@ -1,38 +1,118 @@
...
-        <button className="clear-completed" ></button>
-    </footer>
-</section>
-    </div>
-  );
+class App extends React.Component  {
+    constructor(props) {
+        super(props);
+        this.state = {
+            todos: {
```
Comment (ivadimko) [inline src/App.js]: I'd recommend using an array for items list

File: src/App.js (hunk 1 above)
Comment (ivadimko) [inline src/App.js]: Move methods declaration out of constructor

File: src/App.js [hunk 3]
```diff
@@ -1,38 +1,118 @@
...
+
+        this.getRandomIntInclusive = (min, max) => {
+            min = Math.ceil(min);
+            max = Math.floor(max);
+            return Math.floor(Math.random() * (max - min + 1)) + min;
+          }
+        
+        this.addTodo = (event) => {
+            
+            if (event.key === "Enter") {
```
Comment (ivadimko) [inline src/App.js]: Wrap input in tag `form`. This way you'll have native submit on `enter` press

File: src/App.js [hunk 4]
```diff
@@ -1,38 +1,118 @@
...
+                let newTodo = event.target.value;
+                let newKey = this.generateNewId();
+                event.target.value = '';
```
Comment (ivadimko) [inline src/App.js]: You should avoid this kind of manipulation with DOM.
- Add `value` to state.
- create method `inputChange`
Apply 2-way data binding with input. 
https://reactjs.org/docs/forms.html

File: src/App.js [hunk 5]
```diff
@@ -1,38 +1,118 @@
...
+        }
+
+        this.getRandomIntInclusive = (min, max) => {
+            min = Math.ceil(min);
+            max = Math.floor(max);
+            return Math.floor(Math.random() * (max - min + 1)) + min;
+          }
+        
+        this.addTodo = (event) => {
+
```
Comment (ivadimko) [inline src/App.js]: Prevent adding empty items
image

File: src/App.js [hunk 6]
```diff
@@ -1,38 +1,118 @@
...
+            if (event.key === "Enter") {
+                let newTodo = event.target.value;
+                let newKey = this.generateNewId();
+                event.target.value = '';
+                this.setState({
+                    todos: {
+                        ...this.state.todos,
+                        [newKey]: {
+                            'text': newTodo,
+                            'activ': true
```
Comment (ivadimko) [inline src/App.js]: `active`
Also, you don't need to wrap keys in quotes

File: src/App.js [hunk 7]
```diff
@@ -1,38 +1,118 @@
...
+                        }
+                    }
+                })
+            }    
+        };
+
+        this.checkStatus = (event) => {
+            
+            this.setState((prevState) => {
+                let newData = prevState.todos;
```
Comment (ivadimko) [inline src/App.js]: move calculations out of `setState`, use `const` for unchanged variables

File: src/App.js [hunk 8]
```diff
@@ -1,38 +1,118 @@
...
+                newData[event]['activ'] ? 
+                newData[event]['activ'] = false :
```
Comment (ivadimko) [inline src/App.js]: --- OBJECT MUTATION

Do you remember what happens when you try to assign object to the variable?
```javascript
const a = {
  name: 'Vasya'
  age: 21
}
const b = a;
const c = b;
b.age = 22;
c.name = 'Petya';
console.log(a); // {name: 'Petya', age: 22}
console.log(b); // {name: 'Petya', age: 22}
console.log(c); // {name: 'Petya', age: 22}

Comment (ivadimko) [reply src/App.js]: Let's assume you use array for todos and have prop `id` for each todo item.
```javascript
// first - copy an array
const newData = [...this.state.todos];

// find index of needed item
const itemIndex = newData.findIndex((item) => item.id === id);

// update needed item
newData[itemIndex] = {
  ...newData[itemIndex],
  active: !newData[itemIndex].active
}

// update state with new data
this.setState({
  todos: newData,
})
```

File: src/App.js [hunk 9]
```diff
@@ -1,38 +1,118 @@
...
+                newData[event]['activ'] = true;
+                return {
+                    todos : newData
+                }
+            })
+        }
+
+        this.deleteTask = (task) => {
+
```
Comment (ivadimko) [inline src/App.js]: remove blank line

File: src/App.js [hunk 10]
```diff
@@ -1,38 +1,118 @@
...
+            this.setState((prevState) => {
+                let newData = prevState.todos;
```
Comment (ivadimko) [inline src/App.js]: The same. Object mutation. You should avoid it.
Filtering is easy with array and id:
```javascript
const newData = this.state.todos.filter((item) => item.id !== id)

this.setState({
  items: newData
})
```

File: src/App.js [hunk 11]
```diff
@@ -1,38 +1,118 @@
...
+                return {
+                    todos : newData
+                }
+            })
+        }
+    };
+
+
+    render(){
+        console.log(this.state.todos)
```
Comment (ivadimko) [inline src/App.js]: remove console.log

File: src/App.js [hunk 12]
```diff
@@ -1,38 +1,118 @@
...
+        return (
+            <div className="App">
+                <section className="todoapp">
+    
+                    <header className="header">
+                        <h1>todos</h1>
+                        <input className="new-todo"
+                            placeholder="What needs to be done?" 
+                            onKeyPress={ this.addTodo }
```
Comment (ivadimko) [inline src/App.js]: Implement 2-way data binding

File: src/App.js [hunk 13]
```diff
@@ -1,38 +1,118 @@
...
+                    </header>
+    
+                    <TodoContent 
+                        todoList = { this.state.todos } 
+                        checkActiv = { this.checkStatus }
+                        deleteTask = { this.deleteTask }
+                    />
+    
+                    <footer className="footer" >
+                        <span className="todo-count"><strong>0</strong> items left</span>
```
Comment (ivadimko) [inline src/App.js]: Implement counter

File: src/App.js [hunk 14]
```diff
@@ -1,38 +1,118 @@
...
+                        <ul className="filters">
```
Comment (ivadimko) [inline src/App.js]: Implement filters. You have to refactor your todo-list and make it an array. This way you're supposed to filter items easily.

File: src/App.js [hunk 15]
```diff
@@ -1,38 +1,118 @@
...
+                                <a href="#123" className="selected">All</a>
+                            </li>
+                            <li>
+                                <a href="#11">Active</a>
+                            </li>
+                            <li>
+                                <a href="#11">Completed</a>
+                            </li>
+                        </ul>
+                        <button className="clear-completed" ></button>
```
Comment (ivadimko) [inline src/App.js]: Implement clear button

File: src/components/TodoContent.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+
+function TodoContent(props) {
+    
+    return (
+        
+        <section className="main">
+            <input id="toggle-all" className="toggle-all" type="checkbox" />
```
Comment (ivadimko) [inline src/components/TodoContent.js]: Implement toggle-all logic

File: src/components/TodoContent.js [hunk 17]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+
+function TodoContent(props) {
+
```
Comment (ivadimko) [inline src/components/TodoContent.js]: remove blank lines

File: src/components/TodoContent.js [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+function TodoContent(props) {
+    
+    return (
+        
+        <section className="main">
+            <input id="toggle-all" className="toggle-all" type="checkbox" />
+            <label htmlFor="toggle-all">Mark all as complete</label>
+            <ul className="todo-list">
+               {Object.keys(props.todoList).map((todo) => {
```
Comment (ivadimko) [inline src/components/TodoContent.js]: Refactor items list to array. Look, you use Object.keys which returns an array. It is easier to maintain an array intially.

File: src/components/TodoContent.js [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+                    return (
+                        <li className={ props.todoList[todo]['activ'] ? 'activ' : 'completed' }>
+                            <div className='view'>
+                                <input className="toggle" type="checkbox"
+                                    onClick ={ () => props.checkActiv( todo ) }
```
Comment (ivadimko) [inline src/components/TodoContent.js]:
```jsx
onClick={ () => props.checkActive(todo.id) }
```
Don't pass whole item. You should work with abstract data (like id)

File: src/components/TodoContent.js (hunk 17 above)
Comment (ivadimko) [reply src/components/TodoContent.js]: Use Object destructuring to get rid of repeated word `props` in the code.
```javascript
const {todoList, checkActive, /* other props */} = props;
```

**Pull request merged**